### PR TITLE
[TIR][REFACTOR] ForNode introduce thread binding and remove legacy field

### PR DESCRIPTION
**Pull request by tqchen**
ForNode update
- Remove deprecated device_api.
- Add ThreadBinding for_type.
- Add additional annotations.

More style consistency refactor to make the ForNode
to be consistent with rest of the codebase.

- ForType => ForKind
- Add constant prefix k to enum consts per Google C style
- Introduce ForKind to the python side.

Note that this is mainly a IR data structure update, and compiler/script support for thread binding need to come later

### CODE REVIEW

File: python/tvm/tir/ir_builder.py [hunk 1]
```diff
@@ -249,17 +249,17 @@ def for_range(self, begin, end, name="i", dtype="int32", for_type="serial"):
...
-            elif for_type == "unroll":
-                for_type_id = 3
+            if kind == "serial":
+                kind_id = 0
+            elif kind == "parallel":
+                kind_id = 1
+            elif kind == "vectorize":
+                kind_id = 2
+            elif kind == "unroll":
+                kind_id = 3
```
Comment (Hzfengsy) [inline python/tvm/tir/ir_builder.py:259]: Shall we add thread_binding here?

Comment (tqchen) [reply python/tvm/tir/ir_builder.py:259]: thread binding is not yet supported formally as an API, so we should add it later once we formally land the language support

Comment (jcf94) [inline python/tvm/tir/ir_builder.py:259]:
```suggestion
            if kind == "serial":
                kind_id = ForKind.SERIAL
            elif kind == "parallel":
                kind_id = ForKind.PARALLEL
            elif kind == "vectorize":
                kind_id = ForKind.VECTORIZED
            elif kind == "unroll":
                kind_id = ForKind.UNROLLED
```

Will it be better to also use ForKind as the input of this function?

Comment (tqchen) [reply python/tvm/tir/ir_builder.py:259]: We are trying to be consistent with the old API for now. So let us keep it as it is.

File: src/printer/tir_text_printer.cc [hunk 2]
```diff
@@ -465,27 +465,30 @@ Doc TIRTextPrinter::VisitStmt_(const EvaluateNode* op) {
   return doc;
 }
 
-inline const char* ForType2String(ForType t) {
+inline const char* ForKind2String(ForKind t) {
```
Comment (junrushao1994) [inline src/printer/tir_text_printer.cc:468]: Consider move this to tir/stmt.h?

File: tests/python/unittest/test_tir_transform_loop_partition.py [hunk 3]
```diff
@@ -462,6 +462,7 @@ def test_multilevel_splitting_with_indivisble_factors():
         def visit_stmt(op):
             return isinstance(op, tvm.tir.Max)
 
+        print(lowered_body)
```
Comment (junrushao1994) [inline tests/python/unittest/test_tir_transform_loop_partition.py:465]: Remove this line?

File: src/printer/tir_text_printer.cc (hunk 2 above)
Comment (tqchen) [reply src/printer/tir_text_printer.cc:468]: This case this is specifically used in text printer, so we plan to keep it here for now

File: tests/python/unittest/test_tir_transform_loop_partition.py (hunk 3 above)
Comment (tqchen) [reply tests/python/unittest/test_tir_transform_loop_partition.py:465]: Good catch

**Review by jcf94 — APPROVED** [review]

File: src/printer/tir_text_printer.cc (hunk 2 above)
Comment (junrushao1994) [reply src/printer/tir_text_printer.cc:468]: I thought our convention is do define a stringify method immediately after defining an enum class...

**Review by spectrometerHBH — APPROVED** [review]

**Review by junrushao1994 — APPROVED** [review]

**Comment by junrushao1994** [comment]
Thanks @tqchen @jcf94 @spectrometerHBH! It is merged :-)

**Pull request merged**